connector/src/main/scala/com/datastax/spark/connector/datasource/CassandraCatalog.scala

```diff
@@ -0,0 +1,472 @@
+      }
+    }
+    val columnsToRemove = changes.collect { case remove: DeleteColumn =>
+      checkRemoveNormalColumn(tableMetadata, remove.fieldNames())
+    }
+
+    val columnsToAdd = changes.collect { case add: AddColumn =>
+      (checkColumnName(add.fieldNames()), sparkSqlToJavaDriverType(add.dataType(), protocolVersion))
+    }
+
+    if (propertiesToAdd.nonEmpty) {
+      val setOptionsStatement = propertiesToAdd.foldLeft(
+        SchemaBuilder.alterTable(keyspace, table).asInstanceOf[AlterTableWithOptionsEnd]) { case (alter, prop) =>
+        alter.withOption(prop.property(), parseProperty(prop.value()))
+      }.asCql()
+      logDebug(s"Executing Set Table Properties: $setOptionsStatement")
+      connector.withSessionDo(_.execute(setOptionsStatement))
+    }
+
+    if (columnsToRemove.nonEmpty) {
+      val dropColumnsStatement = SchemaBuilder.alterTable(keyspace, table).dropColumns(columnsToRemove: _*).asCql
+      logDebug(s"Executing Drop Table Columns: $dropColumnsStatement")
+      connector.withSessionDo(_.execute(dropColumnsStatement))
+    }
+
+    if (columnsToAdd.nonEmpty) {
+      val addColumnStatement = columnsToAdd.foldRight(
+        SchemaBuilder.alterTable(keyspace, table).asInstanceOf[AlterTableAddColumn]
+      ) { case ((colName, dataType), alterBuilder) =>
+        alterBuilder.addColumn(colName, dataType)
+      }.asInstanceOf[AlterTableAddColumnEnd].asCql()
+      logDebug(s"Executing Add Table Columns: $addColumnStatement")
+      connector.withSessionDo(_.execute(addColumnStatement))
+    }
+
+    logDebug(s"Table Alteration Complete")
+    loadTable(ident)
+  }
+
+  override def dropTable(ident: Identifier): Boolean = {
+    val tableMeta = getTableMetaData(connector, ident)
+    connector.withSessionDo(_.execute(SchemaBuilder.dropTable(tableMeta.getKeyspace, tableMeta.getName).asCql())).wasApplied()
+  }
+
+  override def renameTable(oldIdent: Identifier, newIdent: Identifier): Unit = {
+    throw new UnsupportedOperationException("Cassandra does not support renaming tables")
+  }
+}
+
+object CassandraCatalog {
+
+  //Add asScala to JavaOptions to make working with the driver a little smoother
+  implicit class ScalaOptionConverter[T](javaOpt: Optional[T]) {
+    def asScala: Option[T] =
+      if (javaOpt.isPresent) Some(javaOpt.get) else None
+  }
+
+
+  val OnlyOneNamespace = "Cassandra only supports a keyspace name of a single level (no periods in keyspace name)"
+
+  //Table Support
+  def getTableMetaData(connector: CassandraConnector, ident: Identifier) = {
+    val namespace = ident.namespace
+    checkNamespace(namespace)
+    val tableMeta = getMetadata(connector)
+      .getKeyspace(fromInternal(namespace.head)).asScala
+      .getOrElse(throw namespaceMissing(getMetadata(connector), namespace))
+      .getTable(fromInternal(ident.name)).asScala
+      .getOrElse(throw tableMissing(getMetadata(connector), namespace, ident.name()))
+    tableMeta
+  }
```

Comment:
Semi-nit: perhaps marginally easier to do with the Java Optional API... at least that way you can avoid the creation of the intermediate Scala options just for a single op:

```scala
  //Table Support
  def getTableMetaData(connector: CassandraConnector, ident: Identifier) = {
    val namespace = ident.namespace
    checkNamespace(namespace)
    getMetadata(connector)
      .getKeyspace(fromInternal(namespace.head))
      .orElseThrow(() => namespaceMissing(getMetadata(connector), namespace))
      .getTable(fromInternal(ident.name))
      .orElseThrow(() => tableMissing(getMetadata(connector), namespace, ident.name()))
  }
```

Similar idea for getKeyspaceMeta() as well.

Oh, and perhaps we should add an explicit return type here (and to getKeyspaceMeta) as well?

Reply:
ah this was before I switched to scala 2.12, you can't write scala 2.11 lambdas and pass them to java functions. But since we are 2.12 we can do this now